control-plane/addMissingIds.cpp

```diff
@@ -0,0 +1,214 @@
+#include "addMissingIds.h"
+
+#include "frontends/p4/evaluator/evaluator.h"
+#include "p4RuntimeAnnotations.h"
+#include "p4RuntimeArchStandard.h"
+
+namespace P4 {
+
+const IR::P4Program* MissingIdAssigner::preorder(IR::P4Program* program) {
+    auto evaluator =  P4::EvaluatorPass(refMap, typeMap);
+    program->apply(evaluator);
+    auto* toplevel = evaluator.getToplevelBlock();
+    CHECK_NULL(toplevel);
+    symbols = ControlPlaneAPI::P4RuntimeSymbolTable::generateSymbols(
+        toplevel->getProgram(), toplevel, refMap, typeMap,
+        archBuilder(refMap, typeMap, toplevel));
+    auto objectVector = IR::Vector<IR::Node>();
+    for (const auto* obj : program->objects) {
+        objectVector.push_back(obj->apply(
```

Comment:
this seems unnecessary - calling the visitor on the program should do exactly this.
You are also allocating a new visitor for each program object, that sounds wasteful.

Reply:
I was getting a segmentation fault for using *this. Let me try again.